### ISSUE TITLE
java.lang.ClassCastException: org.springframework.cloud.sleuth.instrument.async.SpanContinuingTraceRunnable cannot be cast to java.lang.Comparable

### ISSUE THREAD

**Issue by goldthelocks** [spring-cloud/spring-cloud-sleuth#1060]
**Spring cloud version:** 1.3.3

### Details:
I have a custom executor which uses `PriorityBlockingQueue` defined as bean:
```
@Configuration
public class TaskExecutorConfiguration {

	@Bean("sampleExecutor")
	public ThreadPoolExecutor sampleExecutor() {
		Integer processors = getProcessors();
		Integer queueSize = processors * 2;
		return new BlockingTaskExecutor(processors, processors, 0L, TimeUnit.MILLISECONDS,
				new PriorityBlockingQueue<Runnable>(queueSize));
	}

	private Integer getProcessors() {
		return Runtime.getRuntime().availableProcessors();
	}

}
```

This is the `BlockingTaskExecutor` class:
```
public class BlockingTaskExecutor extends ThreadPoolExecutor {

	private final Semaphore semaphore;

	public BlockingTaskExecutor(int corePoolSize, int maximumPoolSize, long keepAliveTime, TimeUnit unit,
			BlockingQueue<Runnable> workQueue) {
		super(corePoolSize, maximumPoolSize, keepAliveTime, unit, workQueue);
		semaphore = new Semaphore(corePoolSize + workQueue.size());
	}	

	@Override
	public void execute(Runnable command) {
		boolean acquired = false;

		do {
			try {
				semaphore.acquire();
				acquired = true;
			} catch (final InterruptedException e) {
				LogUtil.warn(this, "Failed to acquire semaphore");
			}
		} while (!acquired);

		try {
			super.execute(command);
		} catch (final RejectedExecutionException e) {
			semaphore.release();
			throw e;
		}
	}

	@Override
	protected void afterExecute(Runnable r, Throwable t) {
		super.afterExecute(r, t);
		semaphore.release();
	}

}
```

This is how I use it (I am not using `@Async` because of the comparable):

```
    @Autowired
    @Qualifier("sampleExecutor")
    private ThreadPoolExecutor executor;

    public void sampleAsync() {
    	executor.execute(new SampleAsyncTask(request));
    }
```

Then I have `SampleAsyncTask` implementing `Runnable` and `Comparable`:

```
public class SampleAsyncTask implements Runnable, Comparable<SampleAsyncTask> {

	private MySampleRequest request;

	public AsyncSmsSendTask(MySampleRequest request) {
		super();
		this.request = request;
	}

	@Override
	public void run() {
		try {
			sample.processAsync(request);
		} catch (BaseException e) {
			LogUtil.error(this, e.getMessage());
		}
	}

	@Override
	public int compareTo(SampleAsyncTask other) {
		return other.request.getPriority().ordinal() - this.request.getPriority().ordinal();
	}

}
```

### Issue
I was testing my custom executor with `SampleAsyncTask` as its `Runnable` parameter and hit the following exception:
```
Caused by: java.lang.reflect.InvocationTargetException: null
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.cloud.sleuth.instrument.async.ExecutorMethodInterceptor.invoke(ExecutorBeanPostProcessor.java:126)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:673)
	... 75 common frames omitted
Caused by: java.lang.ClassCastException: org.springframework.cloud.sleuth.instrument.async.SpanContinuingTraceRunnable cannot be cast to java.lang.Comparable
	at java.util.concurrent.PriorityBlockingQueue.siftUpComparable(PriorityBlockingQueue.java:357)
	at java.util.concurrent.PriorityBlockingQueue.offer(PriorityBlockingQueue.java:489)
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1371)
	at com.sample.project.executor.BlockingTaskExecutor.execute(BlockingTaskExecutor.java:42)
	at org.springframework.cloud.sleuth.instrument.async.LazyTraceExecutor.execute(LazyTraceExecutor.java:65)
	... 82 common frames omitted
2018-08-16 16:47:53,126 11894 [http-nio-9798-exec-3] ERROR o.a.c.c.C.[.[.[.[dispatcherServlet] - Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is java.lang.reflect.UndeclaredThrowableException] with root cause
java.lang.ClassCastException: org.springframework.cloud.sleuth.instrument.async.SpanContinuingTraceRunnable cannot be cast to java.lang.Comparable
	at java.util.concurrent.PriorityBlockingQueue.siftUpComparable(PriorityBlockingQueue.java:357)
	at java.util.concurrent.PriorityBlockingQueue.offer(PriorityBlockingQueue.java:489)
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1371)
	at com.sample.project.executor.BlockingTaskExecutor.execute(BlockingTaskExecutor.java:42)
	at org.springframework.cloud.sleuth.instrument.async.LazyTraceExecutor.execute(LazyTraceExecutor.java:65)
```

No exceptions if the custom executor (`BlockingTaskExecutor`) is processing the `SampleAsyncTask`. I then tried to use `ArrayBlockingQueue` as the implementation of `BlockingQueue` like so and it's working fine:

```
	@Bean("sampleExecutor")
	public ThreadPoolExecutor sampleExecutor() {
		Integer processors = getProcessors();
		Integer queueSize = processors * 2;
		return new BlockingTaskExecutor(processors, processors, 0L, TimeUnit.MILLISECONDS,
				new ArrayBlockingQueue<Runnable>(queueSize));
	}
```

### Question:
Based from the exception message, does it mean that spring-cloud is using the custom `ThreadPoolExecutor` defined?

If in case this is not an issue, please let me know. Thanks!

**Comment by marcingrzejszczak** [comment]
Have you checked the latest `1.3.4` version?

**Comment by goldthelocks** [comment]
Hi, I tried as suggested but still the same result (exception message).

**Comment by marcingrzejszczak** [comment]
You're the first one to use the `PriorityBlockingQueue` that requires the `Runnable` / `Callable` to be also `Comparable`. The trace wrappers around those classes do not implement `Comparable` - that's why you get that exception. 

So, for now, the workaround is to not use the `PBQ`.

BTW can you check if the same problem is there with Sleuth 2.0.x ?

**Comment by goldthelocks** [comment]
Understood. The workaround I did is to not define my custom executor as `Bean` so when I need to use it, I'll create an instance. But I'm curious if this will be fixed/handled on future release?

Also, unfortunately I cannot test for version 2.0.x because we haven't migrated yet to 2.x.

**Comment by marcingrzejszczak** [comment]
Can you pass the `Comparator` in a constructor, so that you define how `Runnable` / `Callable` gets compared?

**Comment by goldthelocks** [comment]
Hi @marcingrzejszczak , I'm sorry I don't follow. Which class do you mean?

**Comment by marcingrzejszczak** [comment]
`java.util.concurrent.PriorityBlockingQueue#PriorityBlockingQueue(int, java.util.Comparator<? super E>)`. You can pass the `Comparator` and tell it how to compare the objects

**Comment by marcingrzejszczak** [comment]
ping @goldthelocks ?

**Comment by goldthelocks** [comment]
Hello @marcingrzejszczak, sorry for the late revert. We did not proceed with as suggested due to time constraint. But, i'll try it on my local machine once I have time.

**Comment by marcingrzejszczak** [comment]
Ok I'll close until you do

**Comment by adammichalik** [comment]
This is now an issue for me, too with Sleuth 2.0.3. I have:

```java
PriorityBlockingQueue<Runnable> blockingQueue = new PriorityBlockingQueue<>(queueSize, new PriorityFutureComparator());

return new MyThreadPoolExecutor(corePoolSize, maxPoolSize, keepAliveTime, TimeUnit.MILLISECONDS, blockingQueue);
```

I have `MyThreadPoolExecutor extends AbstractExecutorService` because I call `executorService.submit(Callable)`. You can see in the source code of `submit()` that

```java
public <T> Future<T> submit(Callable<T> task) {
    if (task == null) throw new NullPointerException();
    RunnableFuture<T> ftask = newTaskFor(task);
    execute(ftask);
    return ftask;
}
```
first a new `RunnableFuture` is created using `newTaskFor()` and that is submitted to `execute()`. Objects stored in the blocking queue are these futures. For the priority comparator to work, it must get instances that can provide some information about the priority. So `MyThreadPoolExecutor` overrides `newTaskFor(Callable)`:

```java
return new MyPriorityFuture<>(super.newTaskFor(callable), ((MyPriorityCallable) callable).getPriority());
```

Basically at some point we need to be able to get the priority from the callable. But at this point, the callable is already wrapped in `TraceCallable`. I manage to extract the `delegate` with reflection. I don't know if there's anything that Sleuth can do to facilitate this use case, but if at least `TraceCallable` provided a getter for the delegate, that would allow this workaround to work without reflection.

**Comment by Blafasel3** [comment]
Hit this as well, same as @adammichalik. Exposing the delegate would make this problem disappear. 
Another option would be to make TraceRunnable implement Comparable and delegate the call to the the delegate if it implements Comparable.
Any chance of reopening this?
